pgrx/src/datum/array.rs

```diff
@@ -59,15 +54,14 @@ fn with_vec(elems: Array<String>) {
 ```
 */
 pub struct Array<'a, T: FromDatum> {
-    // Remove this field if/when we figure out how to stop using pg_sys::deconstruct_array
     null_slice: NullKind<'a>,
-    elem_layout: Layout,
-    _datum_slice: OnceCell<PallocSlice<pg_sys::Datum>>,
+    slide_impl: ChaChaSlideImpl<T>,
     // Rust drops in FIFO order, drop this last
     raw: Toast<RawArray>,
-    _marker: PhantomData<T>,
 }
 
+type ChaChaSlideImpl<T> = Box<dyn ChaChaSlide<T>>;
```

Comment:
Somewhat surprised that this actually helps the performance that much -- we're replacing static branches with dynamic dispatch. That said, you have the numbers to back it up, so I'm good with it!

Reply:
Well. Same here.  I don’t know if the compiler is able to devirtualize it or not. 

And I couldn’t see a way to get it done with the type system either. Can you?